packages/mobile/routes/login/views/EnterPinView.svelte

```diff
@@ -0,0 +1,155 @@
+    import { validatePinFormat } from '@lib/utils'
+    import { Icon, PinInput, Profile, Text } from 'shared/components'
+    import { onDestroy } from 'svelte'
+
+    let attempts = 0
+    let pinCode = ''
+    let isBusy = false
+    let pinRef
+    let shake = false
+
+    /** Maximum number of consecutive (incorrect) attempts allowed to the user */
+    const MAX_PINCODE_INCORRECT_ATTEMPTS = 3
+
+    /** Waiting time in seconds after which a user should be allowed to enter pin again */
+    const WAITING_TIME_AFTER_MAX_INCORRECT_ATTEMPTS = 30
+
+    let timeRemainingBeforeNextAttempt = WAITING_TIME_AFTER_MAX_INCORRECT_ATTEMPTS
+
+    $: if (needsToAcceptLatestPrivacyPolicy() || needsToAcceptLatestTermsOfService()) {
+        openPopup({
+            type: 'legalUpdate',
+            hideClose: true,
+            preventClose: true,
+        })
+    }
+
+    $: hasReachedMaxAttempts = attempts >= MAX_PINCODE_INCORRECT_ATTEMPTS
+    $: {
+        if (validatePinFormat(pinCode)) {
+            void onSubmitClick()
+        }
+    }
+    $: {
+        if (pinRef && !$popupState.active) {
+            pinRef.focus()
+        }
+    }
+
+    let buttonText = setButtonText(timeRemainingBeforeNextAttempt)
+
+    // @TODO: move to shared lib
+    function setButtonText(time) {
+        return localize('views.login.pleaseWait', { values: { time: time.toString() } })
+    }
+
+    let maxAttemptsTimer = null
+    let shakeTimeout = null
+
+    // @TODO: move to shared lib
+    function countdown(): void {
+        if (!hasReachedMaxAttempts) {
+            return
+        }
+
+        if (timeRemainingBeforeNextAttempt === -1) {
+            clearInterval(maxAttemptsTimer)
+            attempts = 0
+            timeRemainingBeforeNextAttempt = WAITING_TIME_AFTER_MAX_INCORRECT_ATTEMPTS
+            pinRef.resetAndFocus()
+        } else {
+            buttonText = setButtonText(timeRemainingBeforeNextAttempt)
+            timeRemainingBeforeNextAttempt--
+        }
+    }
+
+    // @TODO: move to shared lib
+    async function onSubmitClick(): Promise<void> {
+        if (!hasReachedMaxAttempts) {
+            isBusy = true
+            const isVerified = await Platform.PincodeManager.verify($activeProfile?.id, pinCode)
+            if (isVerified) {
+                void login()
+                $loginRouter.next()
+            } else {
+                shake = true
+                shakeTimeout = setTimeout(() => {
+                    shake = false
+                    isBusy = false
+                    attempts++
+                    if (attempts >= MAX_PINCODE_INCORRECT_ATTEMPTS) {
+                        clearInterval(maxAttemptsTimer)
+                        maxAttemptsTimer = setInterval(countdown, 1000)
+                    } else {
+                        pinRef.resetAndFocus()
+                    }
+                }, 1000)
+            }
+        }
+    }
+
+    // @TODO: move to shared lib
```

Comment:
Same with this, this is a very custom router based thing, so maybe we should keep this kind of logic separate, even if this means code duplication